### DISCUSSION TITLE
Results not filterable on ContentType name

### DISCUSSION THREAD

**Discussion by ACSirus** [microsoft-search/pnp-modern-search#2777] [Q&A] [answered]
I have noticed that there is something wrong with the filtering on content types. I have found loads of files that have a certain content type, but when you filter on that content type name (contenttypeid shows no issues), the file is not shown. See my screen shots below. I got in contact with MS support, but when I use the default search bar in SPO, the problem does not occur, hence they asked me to bring this to your attention. Is there a mistake on my part, or perhaps some kind of bug? For your information I also added the Search Query Tool screen shots :)

![image](https://user-images.githubusercontent.com/77774675/219413579-ac50e5fb-7fb0-4a95-86ca-24e9f6499ff9.png)
![image](https://user-images.githubusercontent.com/77774675/219413840-b8dbacb0-6864-4dd9-a048-694e9a6f2f7c.png)
![image](https://user-images.githubusercontent.com/77774675/219414008-dc58b99d-4ffa-4b91-9694-f49fc2f685b1.png)
![thumbnail_image002](https://user-images.githubusercontent.com/77774675/219414345-79af3e30-01aa-44ff-8aed-8305080b2302.png)
![image004](https://user-images.githubusercontent.com/77774675/219414404-792b5e4f-fc73-4587-ac6f-39f746f1ee94.png)

**Comment by wobba** [comment, marked as answer]
Replace `ContentType` with `SPContentType` and you should be good. ContentType has always had issues as property query since 2013 and often includes "noise", not just the CT string name. And you can just as well add it using KQL in the query template instead of FQL as refiner.

Your user query of `ContentType:Foo` is not equal to FQL with equals, as `:` means contains. So, yes, error is on you 😉 If you add it with KQL and contains instead of FQL and equals you should see the same result. But again, `SPContentType` is the prop to use in queries 🙂

**Reply by ACSirus** [reply]
Thanks so much for the fast and helpful explanation!! This is crucial info when using SharePoint Search for sure!

This indeed helped me retrieve 1.000% more results, but I am still missing 15% - 20% of the total amount of items that should be returned.
The item from my screen shots is still missing though when using equals instead of contains (contains also returns content types that I don't want in my results):
![image](https://user-images.githubusercontent.com/77774675/220058536-616ad6cf-21c5-41be-81cc-07ca1c318dda.png)
![image](https://user-images.githubusercontent.com/77774675/220058845-e7951cbe-872e-4ef8-9122-03b6c76e5878.png)

SSQT shows this for the item in question: Interesting...
![image](https://user-images.githubusercontent.com/77774675/220065988-23c886d7-badb-4eb4-bf4c-44942e3ee201.png)
![image](https://user-images.githubusercontent.com/77774675/220066448-69ab6fff-77ec-4b05-b366-7df4905d87f8.png)

I have sent the same info to our MS support contact as the same is showing in the default SharePoint Search service which is something they can help us with.

Any extra info that you think might be helpful is welcome of course ;)

In any case you have really helped me a lot with this and enriched my knowledge about SP Search. Thanks! 🥇

**Reply by wobba** [reply]
When you say you are missing 15% of items, can you explain? Max page size is 500 I believe and count on refiners is just an estimate.

**Reply by ACSirus** [reply]
When I use the SPContentType:Adviesvraag in the search query it yields me the following results:
![image](https://user-images.githubusercontent.com/77774675/220857622-3946b688-cbe4-45ae-9b0a-9af24389120f.png)
![image](https://user-images.githubusercontent.com/77774675/220859447-ca83aabf-9036-4fbd-bfb5-fc816aeb17bb.png)
(Problem is I don't want all the extra content types that are not exactly "Adviesvraag", "Project" or "Infomap")

When I use SPContentType=Adviesvraag I get:
![image](https://user-images.githubusercontent.com/77774675/220857473-a557f515-2290-4c47-aa9b-a6ce3f51b5ad.png)
![image](https://user-images.githubusercontent.com/77774675/220859549-35fab3c1-8b1f-49dc-918c-921b1038634c.png)

So it's actually more like 10% (2038 of 2245)
Also, it's not just the inaccuracy of the filter in this case I guess as the screen shots above demonstrate that the specific item is not getting retrieved. I asked for support from Microsoft too because SPContentType is null, but ContentType is filled in on the item in the example (see screen shots in my last reply).

So in short for the item in my example:
**Item property**
ContentType: Adviesvraag
SPContentType: <null>

**Search behaviour**
ContentType:Adviesvraag -> item returned in resultset
ContentType=Adviesvraag -> item NOT returned in resultset
SPContentType:Adviesvraag -> item returned in resultset
SPContentType=Adviesvraag -> item NOT returned in resultset

**Reply by kasperbolarsen** [reply]
Can't say that I have seen that issue before,  which might be because I always uses Contenttypeid rather than SPContentype as the stays the same when the name changes.  Perhaps using Contenttypeid could solve the issue?

**Reply by wobba** [reply]
If you want exact match use = and not : `SPContentType=Adviesvraag`

equals vs contains

**Reply by ACSirus** [reply]
> Can't say that I have seen that issue before, which might be because I always uses Contenttypeid rather than SPContentype as the stays the same when the name changes. Perhaps using Contenttypeid could solve the issue?

I tried that too and that seems to work perfectly for retrieving all results, but then users need to also use those id's when filtering, and that won't work in this case.

**Reply by ACSirus** [reply]
> If you want exact match use = and not : `SPContentType=Adviesvraag`
> 
> equals vs contains

Correct. The point I wanted to make was that for some reason an equals doesn't match eventhought it should.
It's with MS Support now anyways, will see if they have any clue what is going on and report back here.

**Reply by kasperbolarsen** [reply]
Regarding the use of contenttypeid vs SPContentType : I use both :-) The ContentTypeID in the query and the SPContentType in the refiners :-)

**Reply by wobba** [reply]
If you have the id, I would always go with the id. In multi-lingual site setups the ct name may have localized names as well, making it less predictable to use.

**Reply by ACSirus** [reply]
I indeed found using the ContentTypeId in the query and SPContentType in the refiners to be the best solution. Thank you all for your help!
